kotest-property/src/commonMain/kotlin/io/kotest/property/Arbitrary.kt

```diff
@@ -0,0 +1,93 @@
+ *
+ * Edge cases are values that are a common source of bugs. For example, a function using ints is
+ * more likely to fail for common edge cases like zero, minus 1, positive 1, [Int.MAX_VALUE] and [Int.MIN_VALUE]
+ * rather than random values like 965489. Therefore, edge cases can be included in sequences
+ * generated by an arbitrary.
+ *
+ * Not all arbitraries will utilize edge cases. For example, if you define an integer generator
+ * using a subset of the number space - say from 100 to 250,000 - then no edge cases are provided.
+ *
+ * Samples are chosen randomly from the sample space and are used to give a greater breadth to
+ * the test cases. For example, in the case of a function using integers, these random values
+ * could be from across the entire integer number line, or could be limited to a subset of ints
+ * such as natural numbers or even numbers.
+ */
+interface Arbitrary<T> : Gen<T> {
+
+   /**
+    * Returns the values that are considered common edge case for the type.
+    *
+    * For example, for Strings this may include the empty string, a string with white space,
+    * a string with unicode, and a string with non-printable characters.
+    *
+    * The result can be empty if for type T there are no common edge cases.
+    *
+    * @return the common edge cases for type T.
+    */
+   fun edgecases(): Iterable<T>
+
+   /**
+    * Returns a sequence of random sample values to be used for testing.
+    *
+    * @param random the [Random] instance to be used for generating values. This random instance is
+    * seeded using the seed provided to the test framework so that tests can be deterministically re-run.
+    * Implementations should honour the random provider whenever possible.
+    *
+    * @return the random test values as instances of [PropertyInput].
+    */
+   fun samples(random: Random): Sequence<PropertyInput<T>>
+
+   override fun generate(random: Random): Sequence<PropertyInput<T>> =
+      edgecases().map { PropertyInput(it) }.asSequence() + samples(random)
+
+   companion object
+}
+
+fun <T, U> Arbitrary<T>.map(f: (T) -> U): Arbitrary<U> = object : Arbitrary<U> {
+   override fun edgecases(): Iterable<U> = this@map.edgecases().map(f)
+   override fun samples(random: Random): Sequence<PropertyInput<U>> =
+      this@map.samples(random).map { it.map(f) }
+}
+
+fun <T> Arbitrary<T>.filter(predicate: (T) -> Boolean): Arbitrary<T> = object : Arbitrary<T> {
+   override fun edgecases(): Iterable<T> = this@filter.edgecases().filter(predicate)
+   override fun samples(random: Random): Sequence<PropertyInput<T>> =
+      this@filter.samples(random).filter { predicate(it.value) }
+}
+
+/**
+ * Returns a new [Arbitrary] where the edge cases of the receiver are replaced with the edge
+ * cases given as input to this function. The samples are unchanged.
+ */
+fun <T> Arbitrary<T>.withEdgeCases(vararg edgecases: T): Arbitrary<T> = withEdgeCases(edgecases.asList())
```

Comment:
To me, `withEdgeCases` mean "Get those edge cases and add these ones to them". Maybe `replaceEdgeCases` might be better here.

What I thought this could be used for would be 
```
Arbitrary.string().withEdgeCases("O'Donnel") //I know that my test specifically failed for a ', but that's not usual for everyone
```

Reply:
But that wouldn't remove default `String` edge cases

Reply:
I've renamed it to setEdgeCases